### PR TITLE
Introduce B2B Notification Template Inheritance

### PR DESCRIPTION
**Pull request by dhaura**
### Purpose
- This PR introduces notification template inheritance across b2b organization hierarchies. Therefore, if no template/s is configured for a given orgnizaition or application, it can inherit the template/s configured in its nearest ancestor organization within its organization hierarchy. If none found within the hierarchy, it will fall back to the system default templates.

### Related Issues
- https://github.com/wso2/product-is/issues/21214

### Related PRs
- https://github.com/wso2-extensions/identity-organization-management/pull/415

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -477,7 +487,7 @@
 
     <properties>
         <!--Identity Governance Version-->
-        <identity.governance.version>1.11.11</identity.governance.version>
+        <identity.governance.version>1.11.18</identity.governance.version>
```
Comment (dhaura) [inline pom.xml:490]: Bump governance version once https://github.com/wso2-extensions/identity-governance/pull/887 is merged.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java [hunk 2]
```diff
@@ -274,37 +257,39 @@ private String getDefaultNotificationLocale(String notificationChannel) {
 
     @Override
     public NotificationTemplate getNotificationTemplate(String notificationChannel, String templateType, String locale,
-                                                        String tenantDomain) throws NotificationTemplateManagerException {
+                                                        String tenantDomain)
+            throws NotificationTemplateManagerException {
 
-        return getNotificationTemplate(notificationChannel, templateType, locale, tenantDomain, null);
+        return getNotificationTemplate(notificationChannel, templateType, locale, tenantDomain, null, true);
```
Comment (AnuradhaSK) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java:263]: shall we comment the reason to set resolve=true in org level email template get method

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java [hunk 3]
```diff
@@ -274,37 +257,39 @@ private String getDefaultNotificationLocale(String notificationChannel) {
...
     }
 
     @Override
     public NotificationTemplate getNotificationTemplate(String notificationChannel, String templateType, String locale,
-                                                        String tenantDomain, String applicationUuid) throws NotificationTemplateManagerException {
+                                                        String tenantDomain, String applicationUuid)
+            throws NotificationTemplateManagerException {
+
+        return getNotificationTemplate(notificationChannel, templateType, locale, tenantDomain, applicationUuid, false);
```
Comment (AnuradhaSK) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java:271]: shall we comment the reason to set resolve=false in app level email template get method

Comment (AnuradhaSK) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java:271]: do the same for other rerouted methods

Comment (AnuradhaSK) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java:271]: Also can't we deprecate the existing methods and point to use the overloaded method with resolve param?

**Review by AnuradhaSK — DISMISSED** [review]

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java (hunk 3 above)
Comment (dhaura) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java]: Addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/291/commits/e631725df3f22df5aad8e687f6a8d483171c8b2b

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java (hunk 2 above)
Comment (dhaura) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java:272]: Addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/291/commits/e631725df3f22df5aad8e687f6a8d483171c8b2b

File: pom.xml (hunk 1 above)
Comment (dhaura) [reply pom.xml]: Bumped with https://github.com/wso2-extensions/identity-event-handler-notification/pull/291/commits/1eb1c1eabcfc569b87195d0f63132736d3ccd55f

**Comment by dhaura** [comment]
Unit tests related to the inheritance behaviour introduced with this PR will be covered by https://github.com/wso2/product-is/issues/21972

**Comment by dhaura** [comment]
- PR builder fails here with the following two test cases.
```
 2024-12-10T11:30:35.7633864Z [ERROR] Tests run: 2937, Failures: 2, Errors: 0, Skipped: 10

 [ERROR] Run 1: EmailTemplatesPositiveTest.testAddEmailTemplateType:192 1 expectation failed.
 Expected status code <201> but was <409>.
 
 [ERROR] Run 2: EmailTemplatesPositiveTest.testAddEmailTemplateType:192 1 expectation failed.
 Expected status code <201> but was <409>.
```
- The reason behind this issue is due to template existence check in `EmailTemplateManagerImpl` returns resolved template existence by default. It says that the template already exists when adding application templates since the org-level template is already available. 
- In order to retrieve the unresolved template existence the new existence check should be called with `resolve` param set to false. This is fix is done to the API layer through https://github.com/wso2/identity-api-server/pull/755/.
- Locally, all notification template management integrations were run with the fix mentioned in the above PR. All tests were passed as mentioned below.
Screenshot 2024-12-10 at 15 23 49

[Template_mgt_TestNG_ Integration Test.pdf](https://github.com/user-attachments/files/18078860/Template_mgt_TestNG_.Integration.Test.pdf)

**Review by AnuradhaSK — APPROVED** [review]

**Comment by dhaura** [comment]
PR builder is successful - https://github.com/wso2/product-is/pull/21999#issuecomment-2531807928

**Pull request merged**